{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "62273b12-c0ad-4eb9-8c1e-b924cb1f2771",
   "metadata": {},
   "outputs": [],
   "source": [
    "import click\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import gffutils\n",
    "import csv\n",
    "from collections import namedtuple\n",
    "import numpy as np\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def match_mRNAs_and_CDSs(mRNA_list, CDS_list):\n",
    "    \"\"\"Join together into a tuple all CDS objects in list and their parent mRNA\"\"\"\n",
    "    matched = []\n",
    "    for mRNA in mRNA_list:\n",
    "        # Start with an empty list of CDSs for each mRNA\n",
    "        all_CDSs_of_mRNA = []\n",
    "        for cds in CDS_list:\n",
    "            if(mRNA.id == cds.attributes.get(\"Parent\")[0]):\n",
    "                all_CDSs_of_mRNA.append(cds)\n",
    "        matched.append((mRNA, all_CDSs_of_mRNA))\n",
    "    return matched\n",
    "\n",
    "def collect_all_mRNAs_and_CDSs(db, gene):\n",
    "    \"\"\"\n",
    "    Gather all mRNA and CDS features of a gene into lists and return \n",
    "    them as a tuple\n",
    "    \"\"\"\n",
    "    mRNAs = []\n",
    "    CDSs = []\n",
    "    for feature in db.children(gene):\n",
    "        if(feature.featuretype == \"mRNA\"):\n",
    "            mRNAs.append(feature)\n",
    "            continue\n",
    "        if(feature.featuretype == \"CDS\"):\n",
    "            CDSs.append(feature)\n",
    "            continue\n",
    "    return (mRNAs, CDSs)\n",
    "\n",
    "def remove_dublicates(list_of_products):\n",
    "    \"\"\"Remove dublicate gene product names from a list\"\"\"\n",
    "    return list(set(list_of_products))\n",
    "\n",
    "def get_products_from_cds_list(cds_list):\n",
    "    \"\"\"Get a list of unique gene product strings from a list of CDS objects\"\"\"\n",
    "    all_products = []\n",
    "    found_products = False\n",
    "    for cds in cds_list:\n",
    "        # If there is a product it is returned inside of a list\n",
    "        product = cds.attributes.get(\"Product\")\n",
    "        if product:\n",
    "            all_products.append(product[0])\n",
    "            found_products = True\n",
    "    if found_products:\n",
    "        return remove_dublicates(all_products)\n",
    "    else:\n",
    "        return list()\n",
    "        \n",
    "def get_IDs_from_cds_list(cds_list):\n",
    "    \"\"\"Get a list of CDS IDs from a list of CDS objects\"\"\"\n",
    "    return [cds.id for cds in cds_list]\n",
    "\n",
    "def create_gene_annotations(db):\n",
    "    \"\"\"Create gene annotions dictionary with all essential info in memory\"\"\"\n",
    "    click.echo(\"Creation of annotions dictionary started...\")\n",
    "    gene_annotations = {}\n",
    "    for gene in db.features_of_type('gene'):\n",
    "        # Gather the gene's all mRNAs and CDSs into own lists\n",
    "        mRNAs, CDSs = collect_all_mRNAs_and_CDSs(db, gene)\n",
    "        # Iterate over CDSs and their parent mRNA and store all relevant info into\n",
    "        # gene_annotation dictionary\n",
    "        for tx_cds in match_mRNAs_and_CDSs(mRNAs, CDSs):        \n",
    "            mRNA = tx_cds[0]\n",
    "            all_CDSs_of_mRNA = tx_cds[1]\n",
    "            cds_products = get_products_from_cds_list(all_CDSs_of_mRNA)\n",
    "            cds_IDs = get_IDs_from_cds_list(all_CDSs_of_mRNA)\n",
    "            # Populate gene_annotations dictionary\n",
    "            if gene.id in gene_annotations:\n",
    "                gene_annotations[gene.id].append({\n",
    "                    \"Gene_name\" : gene.attributes.get(\"Name\"),\n",
    "                    \"Transcript_ID\" : mRNA.id,\n",
    "                    \"Ontology_term\" : mRNA.attributes.get(\"Ontology_term\"),\n",
    "                    \"Dbxref\" : mRNA.attributes.get(\"Dbxref\"),\n",
    "                    \"CDS_ID\" : cds_IDs,\n",
    "                    \"CDS_Product\" : cds_products\n",
    "                    })\n",
    "            else:\n",
    "                gene_annotations[gene.id] = [{\n",
    "                    \"Gene_name\" : gene.attributes.get(\"Name\"),\n",
    "                    \"Transcript_ID\" : mRNA.id,\n",
    "                    \"Ontology_term\" : mRNA.attributes.get(\"Ontology_term\"),\n",
    "                    \"Dbxref\" : mRNA.attributes.get(\"Dbxref\"),\n",
    "                    \"CDS_ID\" : cds_IDs,\n",
    "                    \"CDS_Product\" : cds_products\n",
    "                    }]\n",
    "    click.echo(\"Gene annotions dictionary finished\")\n",
    "    return gene_annotations\n",
    "\n",
    "\n",
    "def parse_refs(ref):\n",
    "    \"\"\"Read an unparsed reference element and return db id\"\"\"\n",
    "    return (\":\").join(ref.split(\":\")[1:])\n",
    "\n",
    "def add_ref(ref_dict, reference_db_name, unparsed_ref_element):\n",
    "    \"\"\"Create a list of reference values\"\"\"\n",
    "    if reference_db_name in ref_dict:\n",
    "        ref_dict[reference_db_name].append(parse_refs(unparsed_ref_element))\n",
    "    else:\n",
    "        ref_dict[reference_db_name] = [parse_refs(unparsed_ref_element)]\n",
    "    return ref_dict\n",
    "\n",
    "def parse_dbxref_list(dbxref_list, all_found):\n",
    "    \"\"\"\n",
    "    Traverse a list of unparsed db references and add them to a dict of lists\n",
    "    Keep also track of which db references didn't exist in the list\n",
    "    \"\"\"\n",
    "    dbxref = {}\n",
    "    for unparsed_reference in dbxref_list:\n",
    "        if (unparsed_reference.startswith(\"UniProtKB\") or unparsed_reference.startswith(\"Swiss-Prot\")):\n",
    "            dbxref = add_ref(dbxref, \"UniProtKB\", unparsed_reference)\n",
    "            all_found['UniProtKB'] = True\n",
    "        elif (unparsed_reference.startswith(\"GeneID\")):\n",
    "            dbxref = add_ref(dbxref, \"GeneID\", unparsed_reference)\n",
    "            all_found['GeneID'] = True\n",
    "        elif (unparsed_reference.startswith(\"KEGG\")):\n",
    "            dbxref = add_ref(dbxref, \"KEGG\", unparsed_reference)\n",
    "            all_found['KEGG'] = True\n",
    "        elif (unparsed_reference.startswith(\"PFAM\")):\n",
    "            dbxref = add_ref(dbxref, \"PFAM\", unparsed_reference)\n",
    "            all_found['PFAM'] = True\n",
    "        elif (unparsed_reference.startswith(\"InterPro\")):\n",
    "            dbxref = add_ref(dbxref, \"InterPro\", unparsed_reference)\n",
    "            all_found['InterPro'] = True\n",
    "        elif (unparsed_reference.startswith(\"EMBL\")):\n",
    "            dbxref = add_ref(dbxref, \"EMBL\", unparsed_reference)\n",
    "            all_found['EMBL'] = True\n",
    "        else:\n",
    "            # If not matched with any, create an unparsed unknown element\n",
    "            dbxref[\"Unknown\"] = unparsed_reference\n",
    "            all_found['Unknown'] = True\n",
    "            click.echo(f\"Unknown reference added: {unparsed_reference}\")\n",
    "    return (dbxref, all_found)\n",
    "\n",
    "def ntuple_to_list(ntuple):\n",
    "    \"\"\"Convert named tuple values to list\"\"\"\n",
    "    return list(ntuple)\n",
    "\n",
    "def parse_Dbxref(dbxref_list):\n",
    "    \"\"\"Parse a list dbxrefs and return a dict of db as key and ids as values\"\"\"\n",
    "    dbxref = {}\n",
    "    # Handle completely missing Dbxref data\n",
    "    if dbxref_list is None:\n",
    "        return {\n",
    "            \"UniProtKB\":[\"\"],\n",
    "            \"GeneID\":[\"\"],\n",
    "            \"KEGG\":[\"\"],\n",
    "            \"PFAM\":[\"\"],\n",
    "            \"InterPro\":[\"\"],\n",
    "            \"EMBL\":[\"\"]\n",
    "            }\n",
    "    \n",
    "    # Keep track of partially missing Dbxref data\n",
    "    all_found = {\n",
    "        \"UniProtKB\" : False,\n",
    "        \"GeneID\" : False,\n",
    "        \"KEGG\" : False,\n",
    "        \"PFAM\" : False,\n",
    "        \"InterPro\" : False,\n",
    "        \"EMBL\" : False,\n",
    "        \"Unknown\" : False\n",
    "    }\n",
    "    # Create dicts for db ids and how partially the references were found\n",
    "    dbxref, all_found = parse_dbxref_list(dbxref_list, all_found)\n",
    "    \n",
    "    # Fill in missing data with empty strings in a list\n",
    "    for db, found in all_found.items():\n",
    "        if not found:\n",
    "            dbxref[db] = [\"\"]\n",
    "    return dbxref\n",
    "\n",
    "def remove_duplicate_values(dbxref_dict):\n",
    "    \"\"\"Go through all db references and remove duplicate db IDs\"\"\"\n",
    "    for key, value in dbxref_dict.items():\n",
    "        dbxref_dict[key] = remove_dublicates(value)\n",
    "    return dbxref_dict\n",
    "\n",
    "def convert_nones_to_list(element):\n",
    "    \"\"\"Convert NoneTypes are converted to empty lists\"\"\"\n",
    "    if element is None:\n",
    "        return []\n",
    "    else:\n",
    "        return element\n",
    "\n",
    "\n",
    "def concat_list(in_list,concat_char = \",\"):\n",
    "    \"\"\"Concatenate all list elements and return a single element list of the results\"\"\"\n",
    "    if isinstance(in_list, list):\n",
    "        return [concat_char.join(in_list)]\n",
    "    else:\n",
    "        return [\"\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "90c594ff-58bb-45e5-8e33-aae513faa229",
   "metadata": {},
   "outputs": [],
   "source": [
    "filename = \"genome.genes.flybasewithcurated.gff3\"\n",
    "database_filename = \"genome.genes.sqlite\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9a42b564-0d46-401b-a30c-d7af4a9b164c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#db = gffutils.create_db(filename, database_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d5ee6b05-264d-495b-8100-29feeeca3b39",
   "metadata": {},
   "outputs": [],
   "source": [
    "#type(db)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c480b678-6a6c-4d9a-817d-5c9baf81f3c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gffutils.FeatureDB(database_filename)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ac1d2fc9-4229-403c-ba5b-5c3dab575f2c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "42206bd5-b247-46de-b090-ef577698d204",
   "metadata": {},
   "outputs": [],
   "source": [
    "from dataclasses import dataclass, field\n",
    "from typing import ClassVar\n",
    "from collections import Counter\n",
    "\n",
    "\n",
    "    \n",
    "    #@staticmethod\n",
    "    #def gather_db_names(cls, dbnames: dict) -> None:\n",
    "    #    cls.all_dbxs.update(dbnames.keys())\n",
    "    \n",
    "    def __post_init__(self):\n",
    "        if self.dbxref_lst:\n",
    "            db_names: list[str] = [get_db_name(db_ref)[0] for db_ref in self.dbxref_lst]\n",
    "            db_name_counts: dict = dict(Counter(db_names))\n",
    "            dbxref_dict = {}\n",
    "            for db_name_ref in self.dbxref_lst:\n",
    "                db_name, db_id = get_db_name(db_name_ref)\n",
    "                num_db_names = db_name_counts.get(db_name)\n",
    "                # Add to dict if only one else update counter dict and add\n",
    "                # unique ending to the dict key\n",
    "                if num_db_names == 1:\n",
    "                    dbxref_dict.update({db_name:db_id})\n",
    "                else:\n",
    "                    db_names_left: int = num_db_names - 1\n",
    "                    db_name_counts.update({db_name:db_names_left})\n",
    "                    new_db_name: str = f\"{db_name}_{num_db_names}\"\n",
    "                    dbxref_dict.update({new_db_name:db_id})\n",
    "            self.dbxref = dbxref_dict\n",
    "            Transcript.all_dbxs.update(dbxref_dict.keys())\n",
    "            #Transcript.all_dbxs.update(dbxref.keys())\n",
    "    \n",
    "    def to_list(self):\n",
    "        \n",
    "        [\n",
    "            self.id,\n",
    "            self.flybase,\n",
    "            self.flybase_revhits,\n",
    "            self.flybase_name,\n",
    "            self.gbue11,\n",
    "            self.gbue11_revhits,\n",
    "            self.orthodb,\n",
    "            self.orthodb_revhits,\n",
    "            self.ontology_term,        \n",
    "        ]\n",
    "    \n",
    "#if index == 0:\n",
    "#    print(element.attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "9fa1662c-2856-4a0c-8dbc-11e95696974d",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcripts = []\n",
    "for index,element in enumerate(db.features_of_type(\"mRNA\")):\n",
    "    transcripts.append(Transcript(\n",
    "        id = element.attributes.get(\"ID\"),\n",
    "        parent = element.attributes.get(\"Parent\"),\n",
    "        flybase = element.attributes.get(\"flybase\"),\n",
    "        flybase_revhits = element.attributes.get(\"flybase_revhits\"),\n",
    "        flybase_name = element.attributes.get(\"flybase_name\"),\n",
    "        gbue11 = element.attributes.get(\"gbue11\"),\n",
    "        gbue11_revhits = element.attributes.get(\"gbue11_revhits\"),\n",
    "        orthodb = element.attributes.get(\"orthodb\"),\n",
    "        orthodb_revhits = element.attributes.get(\"orthodb_revhits\"),\n",
    "        ontology_term = element.attributes.get(\"Ontology_term\"),\n",
    "        dbxref_lst = element.attributes.get(\"Dbxref\"),\n",
    "        #element.attributes.get(\"dbxref\"),\n",
    "    ))\n",
    "    #if index == 19:\n",
    "    #    print(element.attributes)\n",
    "    #    break\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c387460d-f838-4e28-a988-7cc89bc1e96d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Transcript(id=['gbgene5.t1'], parent=['gbgene5'], flybase=['FBgn0033354'], flybase_revhits=None, flybase_name=['FANCI-PA'], gbue11=['GBUE012419-PA'], gbue11_revhits=None, orthodb=['56086_0_003082'], orthodb_revhits=None, ontology_term=['GO:0006281', 'GO:0070182'], dbxref={'FlyBase_Annotation_IDs': 'CG13745-PA', 'FlyBase': 'FBpp0087756', 'FlyMine': 'FBpp0087756', 'GB_protein_2': 'AAF59016', 'GB_protein': 'AAF59016.1', 'modMine': 'FBpp0087756', 'REFSEQ': 'NP_610429', 'UniProt/TrEMBL': 'A1Z7L1'})"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "transcripts[4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "723a7994-eb8b-42fc-b430-98f1742771ff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import csv\n",
    "from collections import namedtuple\n",
    "from dataclasses import dataclass, asdict\n",
    "\n",
    "transcript_annotation_header = [\"id\",\"parent\",\"flybase\",\"flybase_revhits\",\"flybase_name\",\"gbue11\",\"gbue11_revhits\",\"orthodb\",\"orthodb_revhits\",\"ontology_term\"]\n",
    "dbx_refs_sorted = sorted(list(Transcript.all_dbxs))\n",
    "\n",
    "df_data = []\n",
    "header = []\n",
    "diff_exp_data = 'shrinked_padj-filtered.tsv'\n",
    "\n",
    "with open(diff_exp_data, 'r') as reader:\n",
    "    # Handle DE data with column names so it's easier to access each element\n",
    "    diff_exp_reader = csv.reader(reader, delimiter=\"\\t\")\n",
    "    de_header = next(diff_exp_reader)\n",
    "    header = de_header + transcript_annotation_header + dbx_refs_sorted\n",
    "    De_data = namedtuple(\"DE_data\", de_header)\n",
    "    for data in map(De_data._make, diff_exp_reader):\n",
    "        # Capture all data from our current DE data row\n",
    "        current_de_data = list(data)\n",
    "        # Find transcripts that belong to the DE gene\n",
    "        for transcript in transcripts:\n",
    "            if data.Genes == transcript.parent[0]:\n",
    "                # Get each piece of transcript annotation in same order to a list \n",
    "                # so that it can be printed in correct order as annotation to the\n",
    "                # rest of DE data\n",
    "                transcript_annotation = []\n",
    "                for column in transcript_annotation_header:\n",
    "                    column_value = asdict(transcript).get(column)\n",
    "                    # Join all non-None column values to one string\n",
    "                    if column_value:\n",
    "                        transcript_annotation.append(','.join(column_value))\n",
    "                    else:\n",
    "                        # None:s should be written as \"NA\" in the annotation\n",
    "                        transcript_annotation.append(np.nan)\n",
    "                # Fetch all in the DB existing dbx references from the annotation \n",
    "                for dbx_ref in dbx_refs_sorted:\n",
    "                    # If we have a dbx_ref fetch values from it\n",
    "                    if transcript.dbxref:\n",
    "                        dbx_ref_fetched = transcript.dbxref.get(dbx_ref)\n",
    "                        # If the reference value is non-None add it to the annotation\n",
    "                        if dbx_ref_fetched:\n",
    "                            transcript_annotation.append(dbx_ref_fetched)\n",
    "                        else:\n",
    "                            # None:s should be written as \"NA\" in the annotation\n",
    "                            transcript_annotation.append(np.nan)\n",
    "                    else:\n",
    "                        # None:s should be written as \"NA\" in the annotation\n",
    "                        transcript_annotation.append(np.nan)\n",
    "                # Append the data row as a list\n",
    "                df_data.append(current_de_data + transcript_annotation)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "00b89710-7e75-4f3d-b826-621cab8a09e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>id</th>\n",
       "      <th>flybase</th>\n",
       "      <th>flybase_revhits</th>\n",
       "      <th>...</th>\n",
       "      <th>GB_protein_3</th>\n",
       "      <th>GB_protein_4</th>\n",
       "      <th>GB_protein_5</th>\n",
       "      <th>REFSEQ</th>\n",
       "      <th>REFSEQ_2</th>\n",
       "      <th>UniProt/Swiss-Prot</th>\n",
       "      <th>UniProt/TrEMBL</th>\n",
       "      <th>UniProt/TrEMBL_2</th>\n",
       "      <th>modMine</th>\n",
       "      <th>modMine_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gbgene898</td>\n",
       "      <td>50282.2933422622</td>\n",
       "      <td>3.18665147135842</td>\n",
       "      <td>0.375224271502857</td>\n",
       "      <td>8.49281446843104</td>\n",
       "      <td>2.01692141069787e-17</td>\n",
       "      <td>2.71699483235111e-13</td>\n",
       "      <td>gbgene898.t1</td>\n",
       "      <td>FBgn0053126</td>\n",
       "      <td>gbgene898.t1,gbgene11898.t1,gbgene13186.t1,gbg...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_001285560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q8SXR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gbgene18399</td>\n",
       "      <td>4719.41064222729</td>\n",
       "      <td>5.47217063433783</td>\n",
       "      <td>0.660615318009431</td>\n",
       "      <td>8.30708810916151</td>\n",
       "      <td>9.80791214533117e-17</td>\n",
       "      <td>1.32122384509756e-12</td>\n",
       "      <td>gbgene18399.t1</td>\n",
       "      <td>FBgn0000042</td>\n",
       "      <td>gbgene1363.t1,gbgene7972.t1,gbgene9345.t1,gbge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_001284915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P10987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gbgene7672</td>\n",
       "      <td>57572.1989547611</td>\n",
       "      <td>-4.53703656570831</td>\n",
       "      <td>0.952828618088298</td>\n",
       "      <td>-7.38306254249907</td>\n",
       "      <td>1.54689315293287e-13</td>\n",
       "      <td>2.08381976631587e-09</td>\n",
       "      <td>gbgene7672.t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gbgene2487</td>\n",
       "      <td>98.9639623042805</td>\n",
       "      <td>-1.0610887648853</td>\n",
       "      <td>0.868372946758628</td>\n",
       "      <td>-7.24566816785309</td>\n",
       "      <td>4.30312483601809e-13</td>\n",
       "      <td>5.79673946659996e-09</td>\n",
       "      <td>gbgene2487.t1</td>\n",
       "      <td>FBgn0030592</td>\n",
       "      <td>gbgene2487.t1,gbgene14616.t1,gbgene14625.t2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_572981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VY06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0073797</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gbgene15915</td>\n",
       "      <td>36591.4534399807</td>\n",
       "      <td>-4.50257993023928</td>\n",
       "      <td>0.953317548342053</td>\n",
       "      <td>-7.17849580463953</td>\n",
       "      <td>7.04825532781262e-13</td>\n",
       "      <td>9.49470475209638e-09</td>\n",
       "      <td>gbgene15915.t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>gbgene14914</td>\n",
       "      <td>36.3481070818882</td>\n",
       "      <td>2.97653895221461</td>\n",
       "      <td>0.590945529286848</td>\n",
       "      <td>4.99788164664735</td>\n",
       "      <td>5.79635409312971e-07</td>\n",
       "      <td>0.00780826859885504</td>\n",
       "      <td>gbgene14914.t1</td>\n",
       "      <td>FBgn0030589</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_572978</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VY09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0073801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>gbgene6556</td>\n",
       "      <td>276.704688600132</td>\n",
       "      <td>-3.99124773622068</td>\n",
       "      <td>0.806298033937006</td>\n",
       "      <td>-4.98640752957921</td>\n",
       "      <td>6.15123261304436e-07</td>\n",
       "      <td>0.00828632545303206</td>\n",
       "      <td>gbgene6556.t1</td>\n",
       "      <td>FBgn0013812</td>\n",
       "      <td>gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_524424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VDG0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0271744</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>gbgene6556</td>\n",
       "      <td>276.704688600132</td>\n",
       "      <td>-3.99124773622068</td>\n",
       "      <td>0.806298033937006</td>\n",
       "      <td>-4.98640752957921</td>\n",
       "      <td>6.15123261304436e-07</td>\n",
       "      <td>0.00828632545303206</td>\n",
       "      <td>gbgene6556.t2</td>\n",
       "      <td>FBgn0013812</td>\n",
       "      <td>gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_524424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VDG0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0271744</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>gbgene883</td>\n",
       "      <td>135.764437148795</td>\n",
       "      <td>3.09428117848639</td>\n",
       "      <td>0.619886170429192</td>\n",
       "      <td>4.9815493950296</td>\n",
       "      <td>6.30771784056506e-07</td>\n",
       "      <td>0.00849712670302519</td>\n",
       "      <td>gbgene883.t1</td>\n",
       "      <td>FBgn0086687</td>\n",
       "      <td>gbgene883.t1,gbgene1586.t1,gbgene1587.t1,gbgen...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_731711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q7K4Y0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>gbgene10127</td>\n",
       "      <td>71.0677079923662</td>\n",
       "      <td>-3.76706560671461</td>\n",
       "      <td>0.75599509525362</td>\n",
       "      <td>-4.95441088320875</td>\n",
       "      <td>7.25497571021364e-07</td>\n",
       "      <td>0.00977317777922879</td>\n",
       "      <td>gbgene10127.t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Genes          baseMean     log2FoldChange              lfcSE  \\\n",
       "0      gbgene898  50282.2933422622   3.18665147135842  0.375224271502857   \n",
       "1    gbgene18399  4719.41064222729   5.47217063433783  0.660615318009431   \n",
       "2     gbgene7672  57572.1989547611  -4.53703656570831  0.952828618088298   \n",
       "3     gbgene2487  98.9639623042805   -1.0610887648853  0.868372946758628   \n",
       "4    gbgene15915  36591.4534399807  -4.50257993023928  0.953317548342053   \n",
       "..           ...               ...                ...                ...   \n",
       "101  gbgene14914  36.3481070818882   2.97653895221461  0.590945529286848   \n",
       "102   gbgene6556  276.704688600132  -3.99124773622068  0.806298033937006   \n",
       "103   gbgene6556  276.704688600132  -3.99124773622068  0.806298033937006   \n",
       "104    gbgene883  135.764437148795   3.09428117848639  0.619886170429192   \n",
       "105  gbgene10127  71.0677079923662  -3.76706560671461   0.75599509525362   \n",
       "\n",
       "                  stat                pvalue                  padj  \\\n",
       "0     8.49281446843104  2.01692141069787e-17  2.71699483235111e-13   \n",
       "1     8.30708810916151  9.80791214533117e-17  1.32122384509756e-12   \n",
       "2    -7.38306254249907  1.54689315293287e-13  2.08381976631587e-09   \n",
       "3    -7.24566816785309  4.30312483601809e-13  5.79673946659996e-09   \n",
       "4    -7.17849580463953  7.04825532781262e-13  9.49470475209638e-09   \n",
       "..                 ...                   ...                   ...   \n",
       "101   4.99788164664735  5.79635409312971e-07   0.00780826859885504   \n",
       "102  -4.98640752957921  6.15123261304436e-07   0.00828632545303206   \n",
       "103  -4.98640752957921  6.15123261304436e-07   0.00828632545303206   \n",
       "104    4.9815493950296  6.30771784056506e-07   0.00849712670302519   \n",
       "105  -4.95441088320875  7.25497571021364e-07   0.00977317777922879   \n",
       "\n",
       "                 id      flybase  \\\n",
       "0      gbgene898.t1  FBgn0053126   \n",
       "1    gbgene18399.t1  FBgn0000042   \n",
       "2     gbgene7672.t1          NaN   \n",
       "3     gbgene2487.t1  FBgn0030592   \n",
       "4    gbgene15915.t1          NaN   \n",
       "..              ...          ...   \n",
       "101  gbgene14914.t1  FBgn0030589   \n",
       "102   gbgene6556.t1  FBgn0013812   \n",
       "103   gbgene6556.t2  FBgn0013812   \n",
       "104    gbgene883.t1  FBgn0086687   \n",
       "105  gbgene10127.t1          NaN   \n",
       "\n",
       "                                       flybase_revhits  ... GB_protein_3  \\\n",
       "0    gbgene898.t1,gbgene11898.t1,gbgene13186.t1,gbg...  ...          NaN   \n",
       "1    gbgene1363.t1,gbgene7972.t1,gbgene9345.t1,gbge...  ...          NaN   \n",
       "2                                                  NaN  ...          NaN   \n",
       "3          gbgene2487.t1,gbgene14616.t1,gbgene14625.t2  ...          NaN   \n",
       "4                                                  NaN  ...          NaN   \n",
       "..                                                 ...  ...          ...   \n",
       "101                                                NaN  ...          NaN   \n",
       "102  gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...  ...          NaN   \n",
       "103  gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...  ...          NaN   \n",
       "104  gbgene883.t1,gbgene1586.t1,gbgene1587.t1,gbgen...  ...          NaN   \n",
       "105                                                NaN  ...          NaN   \n",
       "\n",
       "    GB_protein_4 GB_protein_5        REFSEQ REFSEQ_2 UniProt/Swiss-Prot  \\\n",
       "0            NaN          NaN  NP_001285560      NaN                NaN   \n",
       "1            NaN          NaN  NP_001284915      NaN             P10987   \n",
       "2            NaN          NaN           NaN      NaN                NaN   \n",
       "3            NaN          NaN     NP_572981      NaN                NaN   \n",
       "4            NaN          NaN           NaN      NaN                NaN   \n",
       "..           ...          ...           ...      ...                ...   \n",
       "101          NaN          NaN     NP_572978      NaN                NaN   \n",
       "102          NaN          NaN     NP_524424      NaN                NaN   \n",
       "103          NaN          NaN     NP_524424      NaN                NaN   \n",
       "104          NaN          NaN     NP_731711      NaN                NaN   \n",
       "105          NaN          NaN           NaN      NaN                NaN   \n",
       "\n",
       "    UniProt/TrEMBL UniProt/TrEMBL_2      modMine modMine_2  \n",
       "0           Q8SXR1              NaN          NaN       NaN  \n",
       "1              NaN              NaN          NaN       NaN  \n",
       "2              NaN              NaN          NaN       NaN  \n",
       "3           Q9VY06              NaN  FBpp0073797       NaN  \n",
       "4              NaN              NaN          NaN       NaN  \n",
       "..             ...              ...          ...       ...  \n",
       "101         Q9VY09              NaN  FBpp0073801       NaN  \n",
       "102         Q9VDG0              NaN  FBpp0271744       NaN  \n",
       "103         Q9VDG0              NaN  FBpp0271744       NaN  \n",
       "104         Q7K4Y0              NaN          NaN       NaN  \n",
       "105            NaN              NaN          NaN       NaN  \n",
       "\n",
       "[106 rows x 34 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.DataFrame(df_data, columns = header)\n",
    "annotated_diff_exp_data = 'shrinked_padj-filtered_annotated.tsv'\n",
    "# Drop all columns where all values are np.nan\n",
    "df = df.dropna(axis=1, how='all')\n",
    "df = df.drop(\"parent\", axis=1)\n",
    "renamings\n",
    "'id':'transcript_ID',\n",
    "'flybase':'FlyBase_ID',\n",
    "'flybase_revhits':'FlyBase_reverse_hits_IDs',\n",
    "'flybase_name':'FlyBase_symbol_name',\n",
    "'ontology_term':'Gene_Ontology_terms',\n",
    "'FlyBase':'FlyBase_reference_ID1',\n",
    "'FlyBase_2':'FlyBase_reference_ID2',\n",
    "'FlyBase_Annotation_IDs':'FlyBase_Annotation_Symbol_ID1',\n",
    "'FlyBase_Annotation_IDs_2':'FlyBase_Annotation_Symbol_ID2',\n",
    "'FlyMine':'FlyMine_ID1',\n",
    "'FlyMine_2':'FlyMine_ID2',\n",
    "'GB_protein':'GB_protein_ID1',\n",
    "'GB_protein_2':'GB_protein_ID2',\n",
    "'GB_protein_3':'GB_protein_ID3',\n",
    "'GB_protein_4':'GB_protein_ID4',\n",
    "'GB_protein_5':'GB_protein_ID5',\n",
    "'REFSEQ':'NCBI_Reference_Sequence_ID1',\n",
    "'REFSEQ_2':'NCBI_Reference_Sequence_ID2',\n",
    "'UniProt/Swiss-Prot':'UniProt/Swiss-Prot_ID',\n",
    "'UniProt/TrEMBL':'UniProt/TrEMBL_ID1',\n",
    "'UniProt/TrEMBL_2':'UniProt/TrEMBL_ID2',\n",
    "}\n",
    "\n",
    "df.rename(columns=renamings)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "c58043eb-903d-4413-b2a2-37dc153cd5fc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_ID</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>FlyBase_ID</th>\n",
       "      <th>FlyBase_reverse_hits_IDs</th>\n",
       "      <th>...</th>\n",
       "      <th>GB_protein_ID3</th>\n",
       "      <th>GB_protein_ID4</th>\n",
       "      <th>GB_protein_ID5</th>\n",
       "      <th>NCBI_Reference_Sequence_ID1</th>\n",
       "      <th>NCBI_Reference_Sequence_ID2</th>\n",
       "      <th>UniProt/Swiss-Prot_ID</th>\n",
       "      <th>UniProt/TrEMBL_ID1</th>\n",
       "      <th>UniProt/TrEMBL_ID2</th>\n",
       "      <th>modMine</th>\n",
       "      <th>modMine_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gbgene898</td>\n",
       "      <td>50282.2933422622</td>\n",
       "      <td>3.18665147135842</td>\n",
       "      <td>0.375224271502857</td>\n",
       "      <td>8.49281446843104</td>\n",
       "      <td>2.01692141069787e-17</td>\n",
       "      <td>2.71699483235111e-13</td>\n",
       "      <td>gbgene898.t1</td>\n",
       "      <td>FBgn0053126</td>\n",
       "      <td>gbgene898.t1,gbgene11898.t1,gbgene13186.t1,gbg...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_001285560</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q8SXR1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gbgene18399</td>\n",
       "      <td>4719.41064222729</td>\n",
       "      <td>5.47217063433783</td>\n",
       "      <td>0.660615318009431</td>\n",
       "      <td>8.30708810916151</td>\n",
       "      <td>9.80791214533117e-17</td>\n",
       "      <td>1.32122384509756e-12</td>\n",
       "      <td>gbgene18399.t1</td>\n",
       "      <td>FBgn0000042</td>\n",
       "      <td>gbgene1363.t1,gbgene7972.t1,gbgene9345.t1,gbge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_001284915</td>\n",
       "      <td>NaN</td>\n",
       "      <td>P10987</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gbgene7672</td>\n",
       "      <td>57572.1989547611</td>\n",
       "      <td>-4.53703656570831</td>\n",
       "      <td>0.952828618088298</td>\n",
       "      <td>-7.38306254249907</td>\n",
       "      <td>1.54689315293287e-13</td>\n",
       "      <td>2.08381976631587e-09</td>\n",
       "      <td>gbgene7672.t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gbgene2487</td>\n",
       "      <td>98.9639623042805</td>\n",
       "      <td>-1.0610887648853</td>\n",
       "      <td>0.868372946758628</td>\n",
       "      <td>-7.24566816785309</td>\n",
       "      <td>4.30312483601809e-13</td>\n",
       "      <td>5.79673946659996e-09</td>\n",
       "      <td>gbgene2487.t1</td>\n",
       "      <td>FBgn0030592</td>\n",
       "      <td>gbgene2487.t1,gbgene14616.t1,gbgene14625.t2</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_572981</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VY06</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0073797</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gbgene15915</td>\n",
       "      <td>36591.4534399807</td>\n",
       "      <td>-4.50257993023928</td>\n",
       "      <td>0.953317548342053</td>\n",
       "      <td>-7.17849580463953</td>\n",
       "      <td>7.04825532781262e-13</td>\n",
       "      <td>9.49470475209638e-09</td>\n",
       "      <td>gbgene15915.t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>101</th>\n",
       "      <td>gbgene14914</td>\n",
       "      <td>36.3481070818882</td>\n",
       "      <td>2.97653895221461</td>\n",
       "      <td>0.590945529286848</td>\n",
       "      <td>4.99788164664735</td>\n",
       "      <td>5.79635409312971e-07</td>\n",
       "      <td>0.00780826859885504</td>\n",
       "      <td>gbgene14914.t1</td>\n",
       "      <td>FBgn0030589</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_572978</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VY09</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0073801</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>102</th>\n",
       "      <td>gbgene6556</td>\n",
       "      <td>276.704688600132</td>\n",
       "      <td>-3.99124773622068</td>\n",
       "      <td>0.806298033937006</td>\n",
       "      <td>-4.98640752957921</td>\n",
       "      <td>6.15123261304436e-07</td>\n",
       "      <td>0.00828632545303206</td>\n",
       "      <td>gbgene6556.t1</td>\n",
       "      <td>FBgn0013812</td>\n",
       "      <td>gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_524424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VDG0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0271744</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>gbgene6556</td>\n",
       "      <td>276.704688600132</td>\n",
       "      <td>-3.99124773622068</td>\n",
       "      <td>0.806298033937006</td>\n",
       "      <td>-4.98640752957921</td>\n",
       "      <td>6.15123261304436e-07</td>\n",
       "      <td>0.00828632545303206</td>\n",
       "      <td>gbgene6556.t2</td>\n",
       "      <td>FBgn0013812</td>\n",
       "      <td>gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_524424</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q9VDG0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>FBpp0271744</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>gbgene883</td>\n",
       "      <td>135.764437148795</td>\n",
       "      <td>3.09428117848639</td>\n",
       "      <td>0.619886170429192</td>\n",
       "      <td>4.9815493950296</td>\n",
       "      <td>6.30771784056506e-07</td>\n",
       "      <td>0.00849712670302519</td>\n",
       "      <td>gbgene883.t1</td>\n",
       "      <td>FBgn0086687</td>\n",
       "      <td>gbgene883.t1,gbgene1586.t1,gbgene1587.t1,gbgen...</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NP_731711</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Q7K4Y0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>gbgene10127</td>\n",
       "      <td>71.0677079923662</td>\n",
       "      <td>-3.76706560671461</td>\n",
       "      <td>0.75599509525362</td>\n",
       "      <td>-4.95441088320875</td>\n",
       "      <td>7.25497571021364e-07</td>\n",
       "      <td>0.00977317777922879</td>\n",
       "      <td>gbgene10127.t1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene_ID          baseMean     log2FoldChange              lfcSE  \\\n",
       "0      gbgene898  50282.2933422622   3.18665147135842  0.375224271502857   \n",
       "1    gbgene18399  4719.41064222729   5.47217063433783  0.660615318009431   \n",
       "2     gbgene7672  57572.1989547611  -4.53703656570831  0.952828618088298   \n",
       "3     gbgene2487  98.9639623042805   -1.0610887648853  0.868372946758628   \n",
       "4    gbgene15915  36591.4534399807  -4.50257993023928  0.953317548342053   \n",
       "..           ...               ...                ...                ...   \n",
       "101  gbgene14914  36.3481070818882   2.97653895221461  0.590945529286848   \n",
       "102   gbgene6556  276.704688600132  -3.99124773622068  0.806298033937006   \n",
       "103   gbgene6556  276.704688600132  -3.99124773622068  0.806298033937006   \n",
       "104    gbgene883  135.764437148795   3.09428117848639  0.619886170429192   \n",
       "105  gbgene10127  71.0677079923662  -3.76706560671461   0.75599509525362   \n",
       "\n",
       "                  stat                pvalue                  padj  \\\n",
       "0     8.49281446843104  2.01692141069787e-17  2.71699483235111e-13   \n",
       "1     8.30708810916151  9.80791214533117e-17  1.32122384509756e-12   \n",
       "2    -7.38306254249907  1.54689315293287e-13  2.08381976631587e-09   \n",
       "3    -7.24566816785309  4.30312483601809e-13  5.79673946659996e-09   \n",
       "4    -7.17849580463953  7.04825532781262e-13  9.49470475209638e-09   \n",
       "..                 ...                   ...                   ...   \n",
       "101   4.99788164664735  5.79635409312971e-07   0.00780826859885504   \n",
       "102  -4.98640752957921  6.15123261304436e-07   0.00828632545303206   \n",
       "103  -4.98640752957921  6.15123261304436e-07   0.00828632545303206   \n",
       "104    4.9815493950296  6.30771784056506e-07   0.00849712670302519   \n",
       "105  -4.95441088320875  7.25497571021364e-07   0.00977317777922879   \n",
       "\n",
       "      transcript_ID   FlyBase_ID  \\\n",
       "0      gbgene898.t1  FBgn0053126   \n",
       "1    gbgene18399.t1  FBgn0000042   \n",
       "2     gbgene7672.t1          NaN   \n",
       "3     gbgene2487.t1  FBgn0030592   \n",
       "4    gbgene15915.t1          NaN   \n",
       "..              ...          ...   \n",
       "101  gbgene14914.t1  FBgn0030589   \n",
       "102   gbgene6556.t1  FBgn0013812   \n",
       "103   gbgene6556.t2  FBgn0013812   \n",
       "104    gbgene883.t1  FBgn0086687   \n",
       "105  gbgene10127.t1          NaN   \n",
       "\n",
       "                              FlyBase_reverse_hits_IDs  ... GB_protein_ID3  \\\n",
       "0    gbgene898.t1,gbgene11898.t1,gbgene13186.t1,gbg...  ...            NaN   \n",
       "1    gbgene1363.t1,gbgene7972.t1,gbgene9345.t1,gbge...  ...            NaN   \n",
       "2                                                  NaN  ...            NaN   \n",
       "3          gbgene2487.t1,gbgene14616.t1,gbgene14625.t2  ...            NaN   \n",
       "4                                                  NaN  ...            NaN   \n",
       "..                                                 ...  ...            ...   \n",
       "101                                                NaN  ...            NaN   \n",
       "102  gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...  ...            NaN   \n",
       "103  gbgene6556.t1,gbgene6556.t2,gbgene7853.t1,gbge...  ...            NaN   \n",
       "104  gbgene883.t1,gbgene1586.t1,gbgene1587.t1,gbgen...  ...            NaN   \n",
       "105                                                NaN  ...            NaN   \n",
       "\n",
       "    GB_protein_ID4 GB_protein_ID5 NCBI_Reference_Sequence_ID1  \\\n",
       "0              NaN            NaN                NP_001285560   \n",
       "1              NaN            NaN                NP_001284915   \n",
       "2              NaN            NaN                         NaN   \n",
       "3              NaN            NaN                   NP_572981   \n",
       "4              NaN            NaN                         NaN   \n",
       "..             ...            ...                         ...   \n",
       "101            NaN            NaN                   NP_572978   \n",
       "102            NaN            NaN                   NP_524424   \n",
       "103            NaN            NaN                   NP_524424   \n",
       "104            NaN            NaN                   NP_731711   \n",
       "105            NaN            NaN                         NaN   \n",
       "\n",
       "    NCBI_Reference_Sequence_ID2 UniProt/Swiss-Prot_ID UniProt/TrEMBL_ID1  \\\n",
       "0                           NaN                   NaN             Q8SXR1   \n",
       "1                           NaN                P10987                NaN   \n",
       "2                           NaN                   NaN                NaN   \n",
       "3                           NaN                   NaN             Q9VY06   \n",
       "4                           NaN                   NaN                NaN   \n",
       "..                          ...                   ...                ...   \n",
       "101                         NaN                   NaN             Q9VY09   \n",
       "102                         NaN                   NaN             Q9VDG0   \n",
       "103                         NaN                   NaN             Q9VDG0   \n",
       "104                         NaN                   NaN             Q7K4Y0   \n",
       "105                         NaN                   NaN                NaN   \n",
       "\n",
       "    UniProt/TrEMBL_ID2      modMine modMine_2  \n",
       "0                  NaN          NaN       NaN  \n",
       "1                  NaN          NaN       NaN  \n",
       "2                  NaN          NaN       NaN  \n",
       "3                  NaN  FBpp0073797       NaN  \n",
       "4                  NaN          NaN       NaN  \n",
       "..                 ...          ...       ...  \n",
       "101                NaN  FBpp0073801       NaN  \n",
       "102                NaN  FBpp0271744       NaN  \n",
       "103                NaN  FBpp0271744       NaN  \n",
       "104                NaN          NaN       NaN  \n",
       "105                NaN          NaN       NaN  \n",
       "\n",
       "[106 rows x 34 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "9714d799-8dee-4dad-9f1a-847b4ada8083",
   "metadata": {},
   "source": [
    "Tehdä nämä kaikki eri osiot funktioiksi ja lisätä riittävät jutut argumenteiksi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "b94661d4-de4f-4b88-ab8f-b0bd8a88a454",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['FlyBase',\n",
       " 'FlyBase_2',\n",
       " 'FlyBase_3',\n",
       " 'FlyBase_4',\n",
       " 'FlyBase_Annotation_IDs',\n",
       " 'FlyBase_Annotation_IDs_2',\n",
       " 'FlyBase_Annotation_IDs_3',\n",
       " 'FlyBase_Annotation_IDs_4',\n",
       " 'FlyMine',\n",
       " 'FlyMine_2',\n",
       " 'GB_protein',\n",
       " 'GB_protein_2',\n",
       " 'GB_protein_3',\n",
       " 'GB_protein_4',\n",
       " 'GB_protein_5',\n",
       " 'GB_protein_6',\n",
       " 'GI',\n",
       " 'REFSEQ',\n",
       " 'REFSEQ_2',\n",
       " 'REFSEQ_3',\n",
       " 'UniProt/Swiss-Prot',\n",
       " 'UniProt/Swiss-Prot_2',\n",
       " 'UniProt/TrEMBL',\n",
       " 'UniProt/TrEMBL_2',\n",
       " 'modMine',\n",
       " 'modMine_2']"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "raw",
   "id": "88e2e347-d5b3-41f8-b8b6-19375c4f7879",
   "metadata": {},
   "source": [
    "transcripts.append(Transcript(\n",
    "        element.attributes.get(\"ID\"),\n",
    "        element.attributes.get(\"Parent\"),\n",
    "        element.attributes.get(\"flybase\"),\n",
    "        element.attributes.get(\"gbue11\"),\n",
    "        element.attributes.get(\"gbue11_revhits\"),\n",
    "        element.attributes.get(\"orthodb\"),\n",
    "        element.attributes.get(\"orthodb_revhits\"),\n",
    "        element.attributes.get(\"Ontology_term\"),\n",
    "        #element.attributes.get(\"dbxref\"),\n",
    "    ))\n",
    "transcripts[2]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8bf7c7c2-adbe-451c-92ea-4e93b95f4057",
   "metadata": {},
   "source": [
    "from collections import Counter\n",
    "db_names: list[str] = [get_db_name(db_ref)[0] for db_ref in dbxref]\n",
    "db_name_counts = dict(Counter(db_names))\n",
    "\n",
    "# Populate dict with unique keys\n",
    "dbxref_dict = {}\n",
    "for db_name_ref in dbxref:\n",
    "    db_name, db_id = get_db_name(db_name_ref)\n",
    "    num_db_names = db_name_counts.get(db_name)\n",
    "    # Add to dict if only one else update counter dict and add\n",
    "    # unique ending to the dict key\n",
    "    if num_db_names == 1:\n",
    "        dbxref_dict.update({db_name:db_id})\n",
    "    else:\n",
    "        db_names_left: int = num_db_names - 1\n",
    "        db_name_counts.update({db_name:db_names_left})\n",
    "        new_db_name: str = f\"{db_name}_{num_db_names}\"\n",
    "        dbxref_dict.update({new_db_name:db_id})"
   ]
  },
  {
   "cell_type": "raw",
   "id": "bda0bf73-0c30-4e97-ab59-c148f8d200d3",
   "metadata": {},
   "source": [
    "from typing import List\n",
    "\n",
    "def gather_db_names(db: gffutils.interface.FeatureDB) -> List[str]:\n",
    "    db_names: set = set()\n",
    "    for index,element in enumerate(db.features_of_type(\"mRNA\")):\n",
    "        # if index == 4:\n",
    "        gb = dict(element.attributes)\n",
    "        dbxref: list = gb.get('Dbxref')\n",
    "        if dbxref:\n",
    "            db_refs = [reference.split(\":\")[0] for reference in dbxref]\n",
    "            db_names.update(db_refs)\n",
    "    return list(db_names)\n",
    "            #print(db_names)\n",
    "\n",
    "            #dbx = {ref.split(\":\")[0]:ref.split(\":\")[1] for ref in dbxref}\n",
    "            #dbx[\"GB_protein\"]"
   ]
  },
  {
   "cell_type": "raw",
   "id": "536ac010-57dc-4096-9a9e-32f955a065cf",
   "metadata": {},
   "source": [
    "db_names = gather_db_names(db)\n",
    "db_names"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b83654e6-1d38-419e-b24c-fec8ca1b9d5b",
   "metadata": {},
   "source": [
    "def pop_by_name(element: str, db_names: List) -> tuple:\n",
    "    if element in db_names: \n",
    "        db_names.remove(element)\n",
    "        return (element, db_names)\n",
    "    else:\n",
    "        return (\"\", db_names)\n",
    "\n",
    "\n",
    "\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "ec99a1e2-5325-4537-adec-d9b358053c81",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FlyBase_Annotation_IDs': 'CG13175-PA',\n",
       " 'FlyBase': 'FBpp0099493',\n",
       " 'FlyMine': 'FBpp0099493',\n",
       " 'GB_protein_2': 'AAF58560',\n",
       " 'GB_protein': 'AAF58560.2',\n",
       " 'modMine': 'FBpp0099493',\n",
       " 'REFSEQ': 'NP_610740',\n",
       " 'UniProt/TrEMBL': 'A1Z8U7'}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dbxref_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "91019823-fb49-4a4a-822b-5f76fe7e76b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ID': ['gbgene5.t1'],\n",
       " 'Parent': ['gbgene5'],\n",
       " 'flybase': ['FBgn0033354'],\n",
       " 'flybase_name': ['FANCI-PA'],\n",
       " 'gbue11': ['GBUE012419-PA'],\n",
       " 'orthodb': ['56086_0_003082'],\n",
       " 'Ontology_term': ['GO:0006281', 'GO:0070182'],\n",
       " 'Dbxref': ['FlyBase_Annotation_IDs:CG13745-PA',\n",
       "  'FlyBase:FBpp0087756',\n",
       "  'FlyMine:FBpp0087756',\n",
       "  'GB_protein:AAF59016',\n",
       "  'GB_protein:AAF59016.1',\n",
       "  'modMine:FBpp0087756',\n",
       "  'REFSEQ:NP_610429',\n",
       "  'UniProt/TrEMBL:A1Z7L1']}"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "5343fd51-e83f-4ca8-809d-da42c77dbc61",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "7a5c110f-5d55-4949-a4f6-f7d1fc8a8d4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = {ref.split(\":\")[0]:ref.split(\":\")[1] for ref in dbxref}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "9a8e9d2f-7a53-4584-b0fe-3ba5777703d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FlyBase_Annotation_IDs': 'CG13175-PA',\n",
       " 'FlyBase': 'FBpp0099493',\n",
       " 'FlyMine': 'FBpp0099493',\n",
       " 'GB_protein': 'AAF58560.2',\n",
       " 'modMine': 'FBpp0099493',\n",
       " 'REFSEQ': 'NP_610740',\n",
       " 'UniProt/TrEMBL': 'A1Z8U7'}"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "16d5ac58-a42e-4eab-9a65-f2b232bc7c7a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'FBpp0099493'"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "a[\"FlyMine\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "54c27ff3-8b39-473e-80e1-8bf14bebe469",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: ['gbgene4.t1']\n",
      "Parent: ['gbgene4']\n",
      "flybase: ['FBgn0053964']\n",
      "flybase_name: ['CG33964-PA']\n",
      "gbue11: ['GBUE012418-PA']\n",
      "orthodb: ['56086_0_003081']\n",
      "Ontology_term: ['GO:0008168']\n",
      "Dbxref: ['FlyBase_Annotation_IDs:CG33964-PA', 'FlyBase:FBpp0099380', 'FlyMine:FBpp0099380', 'GB_protein:AAF58558', 'GB_protein:AAF58558.4', 'modMine:FBpp0099380', 'REFSEQ:NP_001033941', 'UniProt/TrEMBL:Q86NX3']\n"
     ]
    }
   ],
   "source": [
    "print(element.attributes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "7edb7da9-8c01-4c9e-8e52-517b02515c1c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import gffutils\n",
    "db = gffutils.create_db(\"genome.genes.gff3\", \n",
    "                        dbfn='Gb.db', \n",
    "                        force=True, \n",
    "                        keep_order=True,\n",
    "                        merge_strategy='merge', \n",
    "                        sort_attribute_values=True)\n",
    "\n",
    "def match_mRNAs_and_CDSs(mRNA_list, CDS_list):\n",
    "    \"\"\"Join together into a tuple all CDS objects in list and their parent mRNA\"\"\"\n",
    "    matched = []\n",
    "    for mRNA in mRNA_list:\n",
    "        # Start with an empty list of CDSs for each mRNA\n",
    "        all_CDSs_of_mRNA = []\n",
    "        for cds in CDS_list:\n",
    "            if(mRNA.id == cds.attributes.get(\"Parent\")[0]):\n",
    "                all_CDSs_of_mRNA.append(cds)\n",
    "        matched.append((mRNA, all_CDSs_of_mRNA))\n",
    "    return matched\n",
    "\n",
    "def collect_all_mRNAs_and_CDSs(db, gene):\n",
    "    \"\"\"\n",
    "    Gather all mRNA and CDS features of a gene into lists and return \n",
    "    them as a tuple\n",
    "    \"\"\"\n",
    "    mRNAs = []\n",
    "    CDSs = []\n",
    "    for feature in db.children(gene):\n",
    "        if(feature.featuretype == \"mRNA\"):\n",
    "            mRNAs.append(feature)\n",
    "            continue\n",
    "        if(feature.featuretype == \"CDS\"):\n",
    "            CDSs.append(feature)\n",
    "            continue\n",
    "    return (mRNAs, CDSs)\n",
    "\n",
    "def remove_dublicates(list_of_products):\n",
    "    \"\"\"Remove dublicate gene product names from a list\"\"\"\n",
    "    return list(set(list_of_products))\n",
    "\n",
    "def get_products_from_cds_list(cds_list):\n",
    "    \"\"\"Get a list of unique gene product strings from a list of CDS objects\"\"\"\n",
    "    all_products = []\n",
    "    found_products = False\n",
    "    for cds in cds_list:\n",
    "        # If there is a product it is returned inside of a list\n",
    "        product = cds.attributes.get(\"Product\")\n",
    "        if product:\n",
    "            all_products.append(product[0])\n",
    "            found_products = True\n",
    "    if found_products:\n",
    "        return remove_dublicates(all_products)\n",
    "    else:\n",
    "        return list()\n",
    "        \n",
    "def get_IDs_from_cds_list(cds_list):\n",
    "    \"\"\"Get a list of CDS IDs from a list of CDS objects\"\"\"\n",
    "    return [cds.id for cds in cds_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "9278faf5-498b-4141-8095-25ab63ea7d1a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Populate gene dict with all transcript and cds annotations \n",
    "# in a list of dicts, e.g.\n",
    "# {gene1_ID:\n",
    "#[\n",
    "#{'Gene_name': ['HIBN'],\n",
    "# 'Transcript_ID': 'gbgene20309.t1', \n",
    "# 'Ontology_term': ['GO:0005634', 'GO:0061676'], \n",
    "# 'Dbxref': ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901'], \n",
    "# 'CDS_ID': 'gbgene5187.t2.CDS7', \n",
    "# 'CDS_Product': ['Histone-binding protein N1/N2']}, \n",
    "#{'Gene_name': ['HIBN'],\n",
    "# 'Transcript_ID': 'gbgene20309.t2', \n",
    "# 'Ontology_term': ['GO:0005634', 'GO:0061676'], \n",
    "# 'Dbxref': ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901'], \n",
    "# 'CDS_ID': 'gbgene5187.t2.CDS7', \n",
    "# 'CDS_Product': ['Histone-binding protein N1/N2']\n",
    "# ...\n",
    "#],\n",
    "# ...\n",
    "#}\n",
    "gene_annotations = {}\n",
    "for gene in db.features_of_type('gene'):\n",
    "    # Gather the gene's all mRNAs and CDSs into own lists\n",
    "    mRNAs, CDSs = collect_all_mRNAs_and_CDSs(db, gene)\n",
    "    # Iterate over CDSs and their parent mRNA and store all relevant info into\n",
    "    # gene_annotation dictionary\n",
    "    for tx_cds in match_mRNAs_and_CDSs(mRNAs, CDSs):        \n",
    "        mRNA = tx_cds[0]\n",
    "        all_CDSs_of_mRNA = tx_cds[1]\n",
    "        cds_products = get_products_from_cds_list(all_CDSs_of_mRNA)\n",
    "        cds_IDs = get_IDs_from_cds_list(all_CDSs_of_mRNA)\n",
    "        # Populate gene_annotations dictionary\n",
    "        if gene.id in gene_annotations:\n",
    "            gene_annotations[gene.id].append({\n",
    "                \"Gene_name\" : gene.attributes.get(\"Name\"),\n",
    "                \"Transcript_ID\" : mRNA.id,\n",
    "                \"Ontology_term\" : mRNA.attributes.get(\"Ontology_term\"),\n",
    "                \"Dbxref\" : mRNA.attributes.get(\"Dbxref\"),\n",
    "                \"CDS_ID\" : cds_IDs,\n",
    "                \"CDS_Product\" : cds_products\n",
    "                })\n",
    "        else:\n",
    "            gene_annotations[gene.id] = [{\n",
    "                \"Gene_name\" : gene.attributes.get(\"Name\"),\n",
    "                \"Transcript_ID\" : mRNA.id,\n",
    "                \"Ontology_term\" : mRNA.attributes.get(\"Ontology_term\"),\n",
    "                \"Dbxref\" : mRNA.attributes.get(\"Dbxref\"),\n",
    "                \"CDS_ID\" : cds_IDs,\n",
    "                \"CDS_Product\" : cds_products\n",
    "                }]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f97030f5-3c71-4d51-aa07-52cef1adff15",
   "metadata": {},
   "source": [
    "**feature**\n",
    " 'astuple',\n",
    " 'attributes',\n",
    " 'bin',\n",
    " 'calc_bin',\n",
    " 'chrom',\n",
    " 'dialect',\n",
    " 'end',\n",
    " 'extra',\n",
    " 'featuretype',\n",
    " 'file_order',\n",
    " 'frame',\n",
    " 'id',\n",
    " 'keep_order',\n",
    " 'score',\n",
    " 'seqid',\n",
    " 'sequence',\n",
    " 'sort_attribute_values',\n",
    " 'source',\n",
    " 'start',\n",
    " 'stop',\n",
    " 'strand'\n",
    "\n",
    "**db**\n",
    "'add_relation',\n",
    " 'all_features',\n",
    " 'analyze',\n",
    " 'bed12',\n",
    " 'children',\n",
    " 'children_bp',\n",
    " 'conn',\n",
    " 'count_features_of_type',\n",
    " 'create_introns',\n",
    " 'dbfn',\n",
    " 'default_encoding',\n",
    " 'delete',\n",
    " 'dialect',\n",
    " 'directives',\n",
    " 'execute',\n",
    " 'features_of_type',\n",
    " 'featuretypes',\n",
    " 'interfeatures',\n",
    " 'iter_by_parent_childs',\n",
    " 'keep_order',\n",
    " 'merge',\n",
    " 'merge_all',\n",
    " 'method',\n",
    " 'parents',\n",
    " 'pragmas',\n",
    " 'region',\n",
    " 'schema',\n",
    " 'set_pragmas',\n",
    " 'sort_attribute_values',\n",
    " 'update',\n",
    " 'version'\n",
    "\n",
    " **attributes**\n",
    " 'clear',\n",
    " 'get',\n",
    " 'items',\n",
    " 'keys',\n",
    " 'pop',\n",
    " 'popitem',\n",
    " 'setdefault',\n",
    " 'update',\n",
    " 'values'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "4880246c-1ad8-46ba-985b-75fcbe6c2578",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Gene_name': None, 'Transcript_ID': 'gbgene20353.t1', 'Ontology_term': None, 'Dbxref': None, 'CDS_ID': ['gbgene20353.t1.CDS1'], 'CDS_Product': []}]\n"
     ]
    }
   ],
   "source": [
    "print(gene_annotations[gene.id])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "2b73d323-e446-4e7f-9874-81d437053d91",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Gene_name': ['HIBN'], 'Transcript_ID': 'gbgene20309.t1', 'Ontology_term': ['GO:0005634', 'GO:0061676'], 'Dbxref': ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901', 'KEGG:xla:397901', 'PFAM:PF10516', 'PFAM:PF13181', 'InterPro:IPR019544', 'InterPro:IPR013026', 'InterPro:IPR011990', 'InterPro:IPR019734', 'EMBL:X04712'], 'CDS_ID': ['gbgene20309.t1.CDS1'], 'CDS_Product': ['Histone-binding protein N1/N2']}, {'Gene_name': ['HIBN'], 'Transcript_ID': 'gbgene20309.t2', 'Ontology_term': ['GO:0005634', 'GO:0061676'], 'Dbxref': ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901', 'KEGG:xla:397901', 'PFAM:PF10516', 'PFAM:PF13181', 'InterPro:IPR019544', 'InterPro:IPR013026', 'InterPro:IPR011990', 'InterPro:IPR019734', 'EMBL:X04712'], 'CDS_ID': ['gbgene20309.t2.CDS1'], 'CDS_Product': ['Histone-binding protein N1/N2']}]\n"
     ]
    }
   ],
   "source": [
    "print(gene_annotations[\"gbgene20309\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "93140fc1-74d1-45af-902a-1e07d22d497b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ID: ['gbgene20309.t1']\n",
      "Parent: ['gbgene20309']\n",
      "Ontology_term: ['GO:0005634', 'GO:0061676']\n",
      "Dbxref: ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901', 'KEGG:xla:397901', 'PFAM:PF10516', 'PFAM:PF13181', 'InterPro:IPR019544', 'InterPro:IPR013026', 'InterPro:IPR011990', 'InterPro:IPR019734', 'EMBL:X04712']\n"
     ]
    }
   ],
   "source": [
    "gbgene20309_ch = list(db.children(\"gbgene20309\"))\n",
    "for feature in gbgene20309_ch:\n",
    "    if(feature.featuretype == \"mRNA\"):\n",
    "        print(feature.attributes)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "id": "1df8cf4b-5281-4240-824b-82511fbe81f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture cap --no-stderr\n",
    "print(gene_annotations)\n",
    "with open('output.json', 'w') as f:\n",
    "    f.write(cap.stdout)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b38e590c-9497-4234-83c8-7d7fbe93e905",
   "metadata": {},
   "source": [
    "db2 = gffutils.create_db(\"genome.genes.gff3\", \n",
    "                         dbfn='Gb2.db', \n",
    "                         force=True, \n",
    "                         keep_order=True,\n",
    "                         sort_attribute_values=True,\n",
    "                         merge_strategy='merge',\n",
    "                         id_spec=['ID', 'Name'])"
   ]
  },
  {
   "cell_type": "raw",
   "id": "a3ca995a-2ee3-4554-a97a-95d8cfb1286f",
   "metadata": {},
   "source": [
    "all_mRNAs = list(db.features_of_type('mRNA'))\n",
    "all_genes = list(db.features_of_type('gene'))\n",
    "gene_annotations = {}"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d04e43cb-3112-4993-b5e6-905fc768baa0",
   "metadata": {},
   "source": [
    "def check_if_feature_is_CDS(feature):\n",
    "    if(feature.featuretype == \"CDS\"):\n",
    "        return True\n",
    "\n",
    "def get_parent_transcript(db, gene, cds):\n",
    "    for feature in db.children(gene):\n",
    "        if(cds.Parent == feature.id):\n",
    "            return feature\n",
    "def get_"
   ]
  },
  {
   "cell_type": "raw",
   "id": "175cac9d-0f1e-411e-ac35-5c695b7e633b",
   "metadata": {},
   "source": [
    "def retrieve_CDSs_info_from_txID_and_gene(db, tx_ID, gene):\n",
    "    CDS_info_list = []\n",
    "    for feature in db.children(gene):\n",
    "        if(check_if_feature_is_CDS(feature)):\n",
    "            tx = get_parent_transcript(db, gene, feature)\n",
    "                \n",
    "            \n",
    "            \n",
    "            \n",
    "            CDS_info_list.append({\"CDS_ID\" : feature.id, \n",
    "                                  \"Transcript_ID\" : feature.Parent,\n",
    "                                  \"Product\" :feature.attributes.get(\"Product\")})\n",
    "    return CDS_info_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7f319593-3333-4c6e-b6c3-e7140d22a473",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
