{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9eb7cf32-0d2f-4f22-b14e-e24e4ad4d234",
   "metadata": {},
   "outputs": [],
   "source": [
    "import click\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import gffutils\n",
    "import csv\n",
    "from collections import namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d965af23-9aef-450b-9856-1b9f9e883f30",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gff_db(db_file):\n",
    "    \"\"\"Read annotation sqlite database and return FeatureDB instance\"\"\"\n",
    "    click.echo(f\"Reading {db_file} database...\")\n",
    "    return gffutils.FeatureDB(db_file)\n",
    "\n",
    "def match_mRNAs_and_CDSs(mRNA_list, CDS_list):\n",
    "    \"\"\"Join together into a tuple all CDS objects in list and their parent mRNA\"\"\"\n",
    "    matched = []\n",
    "    for mRNA in mRNA_list:\n",
    "        # Start with an empty list of CDSs for each mRNA\n",
    "        all_CDSs_of_mRNA = []\n",
    "        for cds in CDS_list:\n",
    "            if(mRNA.id == cds.attributes.get(\"Parent\")[0]):\n",
    "                all_CDSs_of_mRNA.append(cds)\n",
    "        matched.append((mRNA, all_CDSs_of_mRNA))\n",
    "    return matched\n",
    "\n",
    "def collect_all_mRNAs_and_CDSs(db, gene):\n",
    "    \"\"\"\n",
    "    Gather all mRNA and CDS features of a gene into lists and return \n",
    "    them as a tuple\n",
    "    \"\"\"\n",
    "    mRNAs = []\n",
    "    CDSs = []\n",
    "    for feature in db.children(gene):\n",
    "        if(feature.featuretype == \"mRNA\"):\n",
    "            mRNAs.append(feature)\n",
    "            continue\n",
    "        if(feature.featuretype == \"CDS\"):\n",
    "            CDSs.append(feature)\n",
    "            continue\n",
    "    return (mRNAs, CDSs)\n",
    "\n",
    "def remove_dublicates(list_of_products):\n",
    "    \"\"\"Remove dublicate gene product names from a list\"\"\"\n",
    "    return list(set(list_of_products))\n",
    "\n",
    "def get_products_from_cds_list(cds_list):\n",
    "    \"\"\"Get a list of unique gene product strings from a list of CDS objects\"\"\"\n",
    "    all_products = []\n",
    "    found_products = False\n",
    "    for cds in cds_list:\n",
    "        # If there is a product it is returned inside of a list\n",
    "        product = cds.attributes.get(\"Product\")\n",
    "        if product:\n",
    "            all_products.append(product[0])\n",
    "            found_products = True\n",
    "    if found_products:\n",
    "        return remove_dublicates(all_products)\n",
    "    else:\n",
    "        return list()\n",
    "        \n",
    "def get_IDs_from_cds_list(cds_list):\n",
    "    \"\"\"Get a list of CDS IDs from a list of CDS objects\"\"\"\n",
    "    return [cds.id for cds in cds_list]\n",
    "\n",
    "def create_gene_annotations(db):\n",
    "    \"\"\"Create gene annotions dictionary with all essential info in memory\"\"\"\n",
    "    click.echo(\"Creation of annotions dictionary started...\")\n",
    "    gene_annotations = {}\n",
    "    for gene in db.features_of_type('gene'):\n",
    "        # Gather the gene's all mRNAs and CDSs into own lists\n",
    "        mRNAs, CDSs = collect_all_mRNAs_and_CDSs(db, gene)\n",
    "        # Iterate over CDSs and their parent mRNA and store all relevant info into\n",
    "        # gene_annotation dictionary\n",
    "        for tx_cds in match_mRNAs_and_CDSs(mRNAs, CDSs):        \n",
    "            mRNA = tx_cds[0]\n",
    "            all_CDSs_of_mRNA = tx_cds[1]\n",
    "            cds_products = get_products_from_cds_list(all_CDSs_of_mRNA)\n",
    "            cds_IDs = get_IDs_from_cds_list(all_CDSs_of_mRNA)\n",
    "            # Populate gene_annotations dictionary\n",
    "            if gene.id in gene_annotations:\n",
    "                gene_annotations[gene.id].append({\n",
    "                    \"Gene_name\" : gene.attributes.get(\"Name\"),\n",
    "                    \"Transcript_ID\" : mRNA.id,\n",
    "                    \"Ontology_term\" : mRNA.attributes.get(\"Ontology_term\"),\n",
    "                    \"Dbxref\" : mRNA.attributes.get(\"Dbxref\"),\n",
    "                    \"CDS_ID\" : cds_IDs,\n",
    "                    \"CDS_Product\" : cds_products\n",
    "                    })\n",
    "            else:\n",
    "                gene_annotations[gene.id] = [{\n",
    "                    \"Gene_name\" : gene.attributes.get(\"Name\"),\n",
    "                    \"Transcript_ID\" : mRNA.id,\n",
    "                    \"Ontology_term\" : mRNA.attributes.get(\"Ontology_term\"),\n",
    "                    \"Dbxref\" : mRNA.attributes.get(\"Dbxref\"),\n",
    "                    \"CDS_ID\" : cds_IDs,\n",
    "                    \"CDS_Product\" : cds_products\n",
    "                    }]\n",
    "    click.echo(\"Gene annotions dictionary finished\")\n",
    "    return gene_annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "736ed478-f5bf-49e3-8a5f-ccf244710ddb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# {gene1_ID:\n",
    "# #[\n",
    "# #{'Gene_name': ['HIBN'],\n",
    "# # 'Transcript_ID': 'gbgene20309.t1', \n",
    "# # 'Ontology_term': ['GO:0005634', 'GO:0061676'], \n",
    "# # 'Dbxref': ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901'], \n",
    "# # 'CDS_ID': 'gbgene5187.t2.CDS7', \n",
    "# # 'CDS_Product': ['Histone-binding protein N1/N2']}, \n",
    "# #{'Gene_name': ['HIBN'],\n",
    "# # 'Transcript_ID': 'gbgene20309.t2', \n",
    "# # 'Ontology_term': ['GO:0005634', 'GO:0061676'], \n",
    "# # 'Dbxref': ['UniProtKB/Swiss-Prot:P06180', 'GeneID:397901'], \n",
    "# # 'CDS_ID': 'gbgene5187.t2.CDS7', \n",
    "# # 'CDS_Product': ['Histone-binding protein N1/N2']\n",
    "\n",
    "def ntuple_to_list(ntuple):\n",
    "    return list(ntuple)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2b58e054-07aa-4c60-935b-6b01f7efc7b6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading genome.genes.sqlite database...\n",
      "Creation of annotions dictionary started...\n",
      "Gene annotions dictionary finished\n"
     ]
    }
   ],
   "source": [
    "db = read_gff_db(Path(\"genome.genes.sqlite\"))\n",
    "gene_annotations = create_gene_annotations(db)"
   ]
  },
  {
   "cell_type": "raw",
   "id": "ba8ad93d-9fcf-478a-a65b-4bbf9cf6eb78",
   "metadata": {},
   "source": [
    "%%capture cap --no-stderr\n",
    "print(gene_annotations)\n",
    "with open('output.json', 'w') as f:\n",
    "    f.write(cap.stdout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "d2a5fb99-7baa-4800-9425-426de8704a1f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gbgene898\n",
      "['347']\n",
      "gbgene18399\n",
      "['100337595']\n",
      "gbgene7672\n",
      "['3713']\n",
      "gbgene2487\n",
      "['']\n",
      "gbgene15915\n",
      "['199692']\n",
      "gbgene12445\n",
      "['']\n",
      "gbgene7693\n",
      "['3713']\n",
      "gbgene5076\n",
      "['27385']\n",
      "gbgene6642\n",
      "['658208']\n",
      "gbgene11933\n",
      "['100135492']\n",
      "gbgene9359\n",
      "['27385']\n",
      "gbgene14984\n",
      "['']\n",
      "gbgene15509\n",
      "['']\n",
      "gbgene8521\n",
      "['10847']\n",
      "gbgene8756\n",
      "['']\n",
      "gbgene15500\n",
      "['']\n",
      "gbgene17373\n",
      "['184247']\n",
      "gbgene15499\n",
      "['']\n",
      "gbgene11357\n",
      "['3290499']\n",
      "gbgene12463\n",
      "['396865']\n",
      "gbgene15504\n",
      "['']\n",
      "gbgene5757\n",
      "['27385']\n",
      "gbgene15147\n",
      "['']\n",
      "gbgene1080\n",
      "['']\n",
      "gbgene9384\n",
      "['']\n",
      "gbgene17286\n",
      "['']\n",
      "gbgene7683\n",
      "['3713']\n",
      "gbgene9382\n",
      "['']\n",
      "gbgene16793\n",
      "['']\n",
      "gbgene15534\n",
      "['']\n",
      "gbgene17833\n",
      "['']\n",
      "gbgene13284\n",
      "['57473']\n",
      "gbgene8288\n",
      "['507264']\n",
      "gbgene5761\n",
      "['57473']\n",
      "gbgene7681\n",
      "['3713']\n",
      "gbgene18912\n",
      "['3290499']\n",
      "gbgene19283\n",
      "['']\n",
      "gbgene15749\n",
      "['3290499']\n",
      "gbgene1837\n",
      "['']\n",
      "gbgene7689\n",
      "['']\n",
      "gbgene14858\n",
      "['']\n",
      "gbgene7125\n",
      "['57473']\n",
      "gbgene11967\n",
      "['']\n",
      "gbgene15506\n",
      "['']\n",
      "gbgene11360\n",
      "['3290499']\n",
      "gbgene10081\n",
      "['']\n",
      "gbgene9996\n",
      "['34805']\n",
      "gbgene2941\n",
      "['3290499']\n",
      "gbgene15166\n",
      "['']\n",
      "gbgene15505\n",
      "['']\n",
      "gbgene16792\n",
      "['']\n",
      "gbgene4699\n",
      "['']\n",
      "gbgene312\n",
      "['']\n",
      "gbgene7690\n",
      "['8628578']\n",
      "gbgene11488\n",
      "['']\n",
      "gbgene19926\n",
      "['']\n",
      "gbgene16058\n",
      "['43036']\n",
      "gbgene6061\n",
      "['']\n",
      "gbgene17382\n",
      "['']\n",
      "gbgene7151\n",
      "['']\n",
      "gbgene2781\n",
      "['']\n",
      "gbgene7679\n",
      "['3713']\n",
      "gbgene14859\n",
      "['']\n",
      "gbgene2001\n",
      "['']\n",
      "gbgene7675\n",
      "['']\n",
      "gbgene7682\n",
      "['']\n",
      "gbgene15700\n",
      "['']\n",
      "gbgene5758\n",
      "['27385']\n",
      "gbgene13820\n",
      "['']\n",
      "gbgene9598\n",
      "['']\n",
      "gbgene7691\n",
      "['3713']\n",
      "gbgene15836\n",
      "['23564245']\n",
      "gbgene7120\n",
      "['']\n",
      "gbgene15698\n",
      "['18946']\n",
      "gbgene13460\n",
      "['36880']\n",
      "gbgene14860\n",
      "['']\n",
      "gbgene5882\n",
      "['']\n",
      "gbgene5396\n",
      "['']\n",
      "gbgene5691\n",
      "['40185']\n",
      "gbgene7680\n",
      "['']\n",
      "gbgene1300\n",
      "['']\n",
      "gbgene2760\n",
      "['']\n",
      "gbgene6727\n",
      "['85031']\n",
      "gbgene4711\n",
      "['']\n",
      "gbgene2067\n",
      "['']\n",
      "gbgene9480\n",
      "['']\n",
      "gbgene4712\n",
      "['10915']\n",
      "gbgene2840\n",
      "['']\n",
      "gbgene4691\n",
      "['']\n",
      "gbgene11592\n",
      "['41958']\n",
      "gbgene11962\n",
      "['']\n",
      "gbgene14914\n",
      "['']\n",
      "gbgene6556\n",
      "['']\n",
      "gbgene883\n",
      "['']\n",
      "gbgene10127\n",
      "['']\n",
      "gbgene13992\n",
      "['']\n"
     ]
    }
   ],
   "source": [
    "def parse_refs(ref):\n",
    "    \"\"\"Read an unparsed reference element and return db id\"\"\"\n",
    "    return (\":\").join(ref.split(\":\")[1:])\n",
    "\n",
    "def add_ref(ref_dict, reference_db_name, unparsed_ref_element):\n",
    "    if reference_db_name in ref_dict:\n",
    "        ref_dict[reference_db_name].append(parse_refs(unparsed_ref_element))\n",
    "    else:\n",
    "        ref_dict[reference_db_name] = [parse_refs(unparsed_ref_element)]\n",
    "    return ref_dict\n",
    "        \n",
    "def parse_Dbxref(dbxref_list):\n",
    "    \"\"\"Parse a list dbxrefs and return a dict of db as key and ids as values\"\"\"\n",
    "    dbxref = {}\n",
    "    # Handle completely missing Dbxref data\n",
    "    if dbxref_list is None:\n",
    "        return {\n",
    "            \"UniProtKB\":[\"\"],\n",
    "            \"GeneID\":[\"\"],\n",
    "            \"KEGG\":[\"\"],\n",
    "            \"PFAM\":[\"\"],\n",
    "            \"InterPro\":[\"\"],\n",
    "            \"EMBL\":[\"\"]\n",
    "               }\n",
    "    \n",
    "    # Keep track of partially missing Dbxref data\n",
    "    all_found = {\n",
    "    \"UniProtKB\" : False,\n",
    "    \"GeneID\" : False,\n",
    "    \"KEGG\" : False,\n",
    "    \"PFAM\" : False,\n",
    "    \"InterPro\" : False,\n",
    "    \"EMBL\" : False,\n",
    "    \"Unknown\" : False\n",
    "    }\n",
    "    \n",
    "    for unparse_reference in dbxref_list:\n",
    "        if (unparse_reference.startswith(\"UniProtKB\") or unparse_reference.startswith(\"Swiss-Prot\")):\n",
    "            dbxref = add_ref(dbxref, \"UniProtKB\", unparse_reference)\n",
    "            all_found['UniProtKB'] = True\n",
    "        elif (unparse_reference.startswith(\"GeneID\")):\n",
    "            dbxref = add_ref(dbxref, \"GeneID\", unparse_reference)\n",
    "            all_found['GeneID'] = True\n",
    "        elif (unparse_reference.startswith(\"KEGG\")):\n",
    "            dbxref = add_ref(dbxref, \"KEGG\", unparse_reference)\n",
    "            all_found['KEGG'] = True\n",
    "        elif (unparse_reference.startswith(\"PFAM\")):\n",
    "            dbxref = add_ref(dbxref, \"PFAM\", unparse_reference)\n",
    "            all_found['PFAM'] = True\n",
    "        elif (unparse_reference.startswith(\"InterPro\")):\n",
    "            dbxref = add_ref(dbxref, \"InterPro\", unparse_reference)\n",
    "            all_found['InterPro'] = True\n",
    "        elif (unparse_reference.startswith(\"EMBL\")):\n",
    "            dbxref = add_ref(dbxref, \"EMBL\", unparse_reference)\n",
    "            all_found['EMBL'] = True\n",
    "        else:\n",
    "            # If not matched with any, create an unparsed unknown element\n",
    "            dbxref[\"Unknown\"] = unparse_reference\n",
    "            all_found['Unknown'] = True\n",
    "            print(\"Unknown reference added\"+unparse_reference)\n",
    "    \n",
    "    # Fill in missing data with empty strings in a list\n",
    "    for db, found in all_found.items():\n",
    "        if not found:\n",
    "            dbxref[db] = [\"\"]\n",
    "    return dbxref\n",
    "\n",
    "def remove_duplicate_values(dbxref_dict):\n",
    "    for key, value in dbxref_dict.items():\n",
    "        dbxref_dict[key] = remove_dublicates(value)\n",
    "    return dbxref_dict\n",
    "\n",
    "def check_if_NoneType(element):\n",
    "    if element is None:\n",
    "        return []\n",
    "    else:\n",
    "        return element\n",
    "\n",
    "\n",
    "def concat_list(in_list,concat_char = \",\"):\n",
    "    \"\"\"Concatenate all list elements and return a single element list of the results\"\"\"\n",
    "    if isinstance(in_list, list):\n",
    "        return [concat_char.join(in_list)]\n",
    "    else:\n",
    "        return [\"\"]\n",
    "    \n",
    "\n",
    "with open(Path(\"shrunken_padj-filtered.tsv\"), newline=\"\") as infile, open('output.tsv', 'wt') as out_file:\n",
    "    # Initialise reading and writing tsvs\n",
    "    tsv_writer = csv.writer(out_file, delimiter='\\t')\n",
    "    reader = csv.reader(infile, delimiter=\"\\t\")\n",
    "    \n",
    "    # Write tsv header\n",
    "    header = next(reader)\n",
    "    gff_columns = [\"Gene_name\", \"GeneID\", \"CDS_Products\", \"Ontology_terms\", \"Transcript_IDs\", \"UniProtKB_Swiss-Prot\", \"KEGG\", \"PFAM\", \"InterPro\", \"EMBL\"]\n",
    "    tsv_writer.writerow(header+gff_columns)\n",
    "    \n",
    "    # Read tsv file row-wise\n",
    "    Data = namedtuple(\"Data\", header)  # get names from column headers\n",
    "    for data in map(Data._make, reader):\n",
    "        # Capture DE data into a list\n",
    "        data_row = ntuple_to_list(data)\n",
    "        # Reset all lists/dict for each gene\n",
    "        ontology_terms = []\n",
    "        transcript_IDs = []\n",
    "        CDS_Products = []\n",
    "        dbxref_dict = {}\n",
    "        \n",
    "        # Gather all annotation data \n",
    "        # Gene annotations for each gene ID consist of a list of one or more \n",
    "        # transcripts with all available annotations\n",
    "        for transcript in gene_annotations[data.Genes]:\n",
    "            # gene_name is the same for all the transcripts\n",
    "            gene_name = transcript[\"Gene_name\"]\n",
    "            transcript_IDs.append(transcript[\"Transcript_ID\"])\n",
    "            ontology_terms = ontology_terms + check_if_NoneType(transcript[\"Ontology_term\"])\n",
    "            dbxref_dict = parse_Dbxref(transcript[\"Dbxref\"])\n",
    "            CDS_Products = CDS_Products + transcript[\"CDS_Product\"]\n",
    "        \n",
    "        # Remove dublicates among different transcripts\n",
    "        ontology_terms = remove_dublicates(ontology_terms)\n",
    "        CDS_Products = remove_dublicates(CDS_Products)\n",
    "        dbxref_dict = remove_duplicate_values(dbxref_dict)\n",
    "        \n",
    "        # Write all new data (don't need the DE data) to a row\n",
    "        tsv_writer.writerow([data.Genes]+\n",
    "                            concat_list(gene_name)+\n",
    "                            concat_list(dbxref_dict[\"GeneID\"])+\n",
    "                            concat_list(CDS_Products)+\n",
    "                            concat_list(ontology_terms)+\n",
    "                            concat_list(transcript_IDs)+\n",
    "                            concat_list(dbxref_dict[\"UniProtKB\"])+\n",
    "                            concat_list(dbxref_dict[\"KEGG\"])+\n",
    "                            concat_list(dbxref_dict[\"PFAM\"])+\n",
    "                            concat_list(dbxref_dict[\"InterPro\"])+\n",
    "                            concat_list(dbxref_dict[\"EMBL\"]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "90c41527-a2d6-4f66-a53f-9fb07a6a42f8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ontology_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "9bad2c0c-2c9f-4895-9df8-11ff366f4106",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(['APOD_0'],\n",
       " {'UniProtKB': ['P05090'],\n",
       "  'GeneID': ['347'],\n",
       "  'KEGG': ['hsa:347'],\n",
       "  'PFAM': ['PF08212'],\n",
       "  'InterPro': ['IPR026222',\n",
       "   'IPR002969',\n",
       "   'IPR012674',\n",
       "   'IPR022272',\n",
       "   'IPR022271',\n",
       "   'IPR000566'],\n",
       "  'EMBL': ['J02611',\n",
       "   'M16696',\n",
       "   'CR456838',\n",
       "   'AK312090',\n",
       "   'BC007402',\n",
       "   'CH471052',\n",
       "   'BT019860',\n",
       "   'CR541773',\n",
       "   'S80440',\n",
       "   'BT019861']},\n",
       " ['Apolipoprotein D'],\n",
       " ['gbgene898.t1'])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name, dbxref_dict, CDS_Products, transcript_IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "b9d401dc-93df-4766-aa12-0e9e6fc020e9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AL162596'"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "parse_refs(\"EMBL:AL162596\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "588ad0b5-87c5-4898-aafc-af3af9fa3ce4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'Gene_name': ['IVL_13'], 'Transcript_ID': 'gbgene7693.t1', 'Ontology_term': ['GO:0005813', 'GO:0001533', 'GO:0005737', 'GO:0005829', 'GO:0070062', 'GO:0016604', 'GO:0070268', 'GO:0018153', 'GO:0030216', 'GO:0018149', 'GO:0010224'], 'Dbxref': ['UniProtKB/Swiss-Prot:P07476', 'GeneID:3713', 'KEGG:hsa:3713', 'PFAM:PF00904', 'PFAM:PF10583', 'InterPro:IPR002360', 'InterPro:IPR019743', 'InterPro:IPR019571', 'InterPro:IPR000354', 'EMBL:M13903', 'EMBL:AL162596', 'EMBL:BC046391'], 'CDS_ID': ['gbgene7693.t1.CDS1', 'gbgene7693.t1.CDS2'], 'CDS_Product': ['Involucrin']}, {'Gene_name': ['IVL_13'], 'Transcript_ID': 'gbgene7693.t2', 'Ontology_term': ['GO:0005813', 'GO:0001533', 'GO:0005737', 'GO:0005829', 'GO:0070062', 'GO:0016604', 'GO:0070268', 'GO:0018153', 'GO:0030216', 'GO:0018149', 'GO:0010224'], 'Dbxref': ['UniProtKB/Swiss-Prot:P07476', 'GeneID:3713', 'KEGG:hsa:3713', 'PFAM:PF00904', 'PFAM:PF10583', 'InterPro:IPR002360', 'InterPro:IPR019743', 'InterPro:IPR019571', 'InterPro:IPR000354', 'EMBL:M13903', 'EMBL:AL162596', 'EMBL:BC046391'], 'CDS_ID': ['gbgene7693.t2.CDS1', 'gbgene7693.t2.CDS2'], 'CDS_Product': ['Involucrin']}]\n"
     ]
    }
   ],
   "source": [
    "print(gene_annotations['gbgene7693'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "887ca0db-4dc9-46ac-9365-2ca87050d040",
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate(gene_annotations, unannotated_DE_results):\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
