{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d3be9f20-4088-4084-81d8-852fb3ceb9f0",
   "metadata": {},
   "source": [
    "This notebook uses gffutils more directly to handle annotations."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4c7614a-9aa4-4cc3-a8d9-ba9e699330bd",
   "metadata": {},
   "outputs": [],
   "source": [
    "import click\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "import gffutils\n",
    "import csv\n",
    "from collections import namedtuple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "4cd80ed2-0619-4527-8d2c-6b503f77ba11",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_gff_db(db_file):\n",
    "    \"\"\"Read annotation sqlite database and return FeatureDB instance\"\"\"\n",
    "    click.echo(f\"Reading {db_file} database...\")\n",
    "    return gffutils.FeatureDB(db_file)\n",
    "\n",
    "def remove_dublicates(list_of_GOs):\n",
    "    \"\"\"Remove dublicate GO terms from a list\"\"\"\n",
    "    #click.echo(\"Removing duplicate GO-terms...\")\n",
    "    return list(set(list_of_GOs))\n",
    "\n",
    "def join_dict_value_lists(all_genes):\n",
    "    \"\"\"\n",
    "    Iterate over all list of lists in dict values and update the value to\n",
    "    one concatenated list\n",
    "    \"\"\"\n",
    "    click.echo(\"Joining list of duplicated GO-terms into one and removing duplicates...\")\n",
    "    for key,value in all_genes.items():\n",
    "        if value is not None:\n",
    "            all_genes[key] = remove_dublicates(sum(value, []))\n",
    "    click.echo(\"Lists concatenated and GO-terms deduplicated\")\n",
    "    return all_genes\n",
    "\n",
    "def collect_all_mRNAs(db, gene):\n",
    "    \"\"\"Gather all mRNAs of a gene into lists and return them as a list\"\"\"\n",
    "    #click.echo(\"Collecting all mRNAs from genes...\")\n",
    "    mRNAs = []\n",
    "    for feature in db.children(gene):\n",
    "        if(feature.featuretype == \"mRNA\"):\n",
    "            mRNAs.append(feature)\n",
    "    #click.echo(\"All mRNAs from genes collected\")\n",
    "    return mRNAs\n",
    "\n",
    "def populate_GO_dict(db):\n",
    "    \"\"\"\n",
    "    Read FeatureDB and return a dict of gene IDs as keys and unique GO terms as\n",
    "    values\n",
    "    \"\"\"\n",
    "    click.echo(\"Populating GO-terms dictionary...\")\n",
    "    all_genes = {}\n",
    "    for gene in db.features_of_type('gene'):\n",
    "        for mRNA in collect_all_mRNAs(db, gene):\n",
    "            GO_terms = mRNA.attributes.get(\"Ontology_term\")\n",
    "            if GO_terms is None:\n",
    "                GO_terms = []\n",
    "            if gene.id in all_genes:\n",
    "                all_genes[gene.id].append(GO_terms)\n",
    "            else:\n",
    "                all_genes[gene.id] = [GO_terms]\n",
    "    click.echo(\"GO-terms dictionary populated\")\n",
    "    return join_dict_value_lists(all_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "7fc99cd4-bea2-414d-8572-ba22a3037bfe",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading genome.genes.sqlite database...\n"
     ]
    }
   ],
   "source": [
    "db = read_gff_db(Path(\"genome.genes.sqlite\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "444fcbc1-28ac-4cd4-a6e7-2815340c5598",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gbscaf00001\t.\tmRNA\t10755\t50930\t.\t+\t.\tID=gbgene1.t1;Parent=gbgene1;\n"
     ]
    }
   ],
   "source": [
    "for mrna in db.features_of_type('mRNA'):\n",
    "    print(mrna)\n",
    "    break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8a856c42-e5c5-4e66-9165-2de2b3bf6025",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('ID', ['gbgene1.t1'])\n",
      "('Parent', ['gbgene1'])\n"
     ]
    }
   ],
   "source": [
    "for attribute in mrna.attributes.items():\n",
    "    print(attribute)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "9f3609c5-0992-4301-a7d4-75ad53ad288a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on Feature in module gffutils.feature object:\n",
      "\n",
      "class Feature(builtins.object)\n",
      " |  Feature(seqid='.', source='.', featuretype='.', start='.', end='.', score='.', strand='.', frame='.', attributes=None, extra=None, bin=None, id=None, dialect=None, file_order=None, keep_order=False, sort_attribute_values=False)\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __eq__(self, other)\n",
      " |      Return self==value.\n",
      " |  \n",
      " |  __getitem__(self, key)\n",
      " |  \n",
      " |  __hash__(self)\n",
      " |      Return hash(self).\n",
      " |  \n",
      " |  __init__(self, seqid='.', source='.', featuretype='.', start='.', end='.', score='.', strand='.', frame='.', attributes=None, extra=None, bin=None, id=None, dialect=None, file_order=None, keep_order=False, sort_attribute_values=False)\n",
      " |      Represents a feature from the database.\n",
      " |      \n",
      " |      Usually you won't want to use this directly, since it has various\n",
      " |      implementation details needed for operating in the context of FeatureDB\n",
      " |      objects.  Instead, try the :func:`gffutils.feature.feature_from_line`\n",
      " |      function.\n",
      " |      \n",
      " |      When printed, reproduces the original line from the file as faithfully\n",
      " |      as possible using `dialect`.\n",
      " |      \n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      \n",
      " |      seqid : string\n",
      " |          Name of the sequence (often chromosome)\n",
      " |      \n",
      " |      source : string\n",
      " |          Source of the feature; typically the originating database or\n",
      " |          program that predicted the feature\n",
      " |      \n",
      " |      featuretype : string\n",
      " |          Type of feature.  For example \"gene\", \"exon\", \"TSS\", etc\n",
      " |      \n",
      " |      start, end : int or \".\"\n",
      " |          1-based coordinates; start must be <= end.  If \".\" (the default\n",
      " |          placeholder for GFF files), then the corresponding attribute will\n",
      " |          be None.\n",
      " |      \n",
      " |      score : string\n",
      " |          Stored as a string.\n",
      " |      \n",
      " |      strand : \"+\" | \"-\" | \".\"\n",
      " |          Strand of the feature; \".\" when strand is not relevant.\n",
      " |      \n",
      " |      frame : \"0\" | \"1\" | \"2\"\n",
      " |          Coding frame.  0 means in-frame; 1 means there is one extra base at\n",
      " |          the beginning, so the first codon starts at the second base;\n",
      " |          2 means two extra bases at the beginning.  Interpretation is strand\n",
      " |          specific; \"beginning\" for a minus-strand feature is at the end\n",
      " |          coordinate.\n",
      " |      \n",
      " |      attributes : string or dict\n",
      " |          If a string, first assume it is serialized JSON; if this fails then\n",
      " |          assume it's the original key/vals string.  If it's a dictionary\n",
      " |          already, then use as-is.\n",
      " |      \n",
      " |          The end result is that this instance's `attributes` attribute will\n",
      " |          always be a dictionary.\n",
      " |      \n",
      " |          Upon printing, the attributes will be reconstructed based on this\n",
      " |          dictionary and the dialect -- except if the original attributes\n",
      " |          string was provided, in which case that will be used directly.\n",
      " |      \n",
      " |          Notes on encoding/decoding: the only time unquoting\n",
      " |          (e.g., \"%2C\" becomes \",\") happens is if `attributes` is a string\n",
      " |          and if `settings.ignore_url_escape_characters = False`. If dict or\n",
      " |          JSON, the contents are used as-is.\n",
      " |      \n",
      " |          Similarly, the only time characters are quoted (\",\" becomes \"%2C\")\n",
      " |          is when the feature is printed (`__str__` method).\n",
      " |      \n",
      " |      extra : string or list\n",
      " |          Additional fields after the canonical 9 fields for GFF/GTF.\n",
      " |      \n",
      " |          If a string, then first assume it's serialized JSON; if this fails\n",
      " |          then assume it's a tab-delimited string of additional fields.  If\n",
      " |          it's a list already, then use as-is.\n",
      " |      \n",
      " |      bin : int\n",
      " |          UCSC genomic bin. If None, will be created based on provided\n",
      " |          start/end; if start or end is \".\" then bin will be None.\n",
      " |      \n",
      " |      id : None or string\n",
      " |          Database-specific primary key for this feature.  The only time this\n",
      " |          should not be None is if this feature is coming from a database, in\n",
      " |          which case it will be filled in automatically.\n",
      " |      \n",
      " |      dialect : dict or None\n",
      " |      \n",
      " |          The dialect to use when reconstructing attribute strings; defaults\n",
      " |          to the GFF3 spec.  :class:`FeatureDB` objects will automatically\n",
      " |          attach the dialect from the original file.\n",
      " |      \n",
      " |      file_order : int\n",
      " |          This is the `rowid` special field used in a sqlite3 database; this\n",
      " |          is provided by FeatureDB.\n",
      " |      \n",
      " |      keep_order : bool\n",
      " |          If True, then the attributes in the printed string will be in the\n",
      " |          order specified in the dialect.  Disabled by default, since this\n",
      " |          sorting step is time-consuming over many features.\n",
      " |      \n",
      " |      sort_attribute_values : bool\n",
      " |          If True, then the values of each attribute will be sorted when the\n",
      " |          feature is printed.  Mostly useful for testing, where the order is\n",
      " |          important for checking against expected values. Disabled by\n",
      " |          default, since it can be time-consuming over many features.\n",
      " |  \n",
      " |  __len__(self)\n",
      " |  \n",
      " |  __ne__(self, other)\n",
      " |      Return self!=value.\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |      Return repr(self).\n",
      " |  \n",
      " |  __setitem__(self, key, value)\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  __unicode__(self)\n",
      " |  \n",
      " |  astuple(self, encoding=None)\n",
      " |      Return a tuple suitable for import into a database.\n",
      " |      \n",
      " |      Attributes field and extra field jsonified into strings. The order of\n",
      " |      fields is such that they can be supplied as arguments for the query\n",
      " |      defined in :attr:`gffutils.constants._INSERT`.\n",
      " |      \n",
      " |      If `encoding` is not None, then convert string fields to unicode using\n",
      " |      the provided encoding.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      Tuple\n",
      " |  \n",
      " |  calc_bin(self, _bin=None)\n",
      " |      Calculate the smallest UCSC genomic bin that will contain this feature.\n",
      " |  \n",
      " |  sequence(self, fasta, use_strand=True)\n",
      " |      Retrieves the sequence of this feature as a string.\n",
      " |      \n",
      " |      Uses the pyfaidx package.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      \n",
      " |      fasta : str\n",
      " |          If str, then it's a FASTA-format filename; otherwise assume it's\n",
      " |          a pyfaidx.Fasta object.\n",
      " |      \n",
      " |      use_strand : bool\n",
      " |          If True (default), the sequence returned will be\n",
      " |          reverse-complemented for minus-strand features.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      string\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  chrom\n",
      " |  \n",
      " |  stop\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(mrna)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
